In [2]:
%matplotlib inline
import numpy as np
import pandas as pd

import viz


##### 1. How likely is it that you roll doubles when rolling two dice?

In [13]:
n_trials = nrows = 10_000
n_dice = ncols = 2

rolls = np.random.choice([1,2,3,4,5,6], n_trials * n_dice).reshape(nrows, ncols)

In [32]:
rolls

array([[6, 2],
       [1, 3],
       [2, 5],
       ...,
       [6, 4],
       [3, 5],
       [3, 3]])

In [45]:
df = pd.DataFrame(rolls)

In [46]:
df

0  1
0     6  2
1     1  3
2     2  5
3     4  1
4     4  3
...  .. ..
9995  6  3
9996  3  3
9997  6  4
9998  3  5
9999  3  3

[10000 rows x 2 columns]

In [47]:
(df[0] == df[1]).mean()

0.1644

In [50]:
df[(df[0]) == (df[1])]

0  1
5     3  3
6     1  1
9     2  2
18    3  3
20    3  3
...  .. ..
9980  3  3
9982  6  6
9984  4  4
9996  3  3
9999  3  3

[1644 rows x 2 columns]

### The probability of rolling doubles with two dice is 0.1644 or 16.44%

##### 

##### 2. If you flip 8 coins, 
a. what is the probability of getting exactly 3 heads?

b. What is the probability of getting more than 3 heads?


In [23]:
n_trials = n_cols = 10_000

n_coins = n_rows = 8

flips = np.random.choice([1, 0], n_trials * n_coins).reshape(n_cols, n_rows)

In [24]:
flips

array([[0, 1, 1, ..., 1, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       [1, 0, 1, ..., 0, 0, 0],
       ...,
       [1, 0, 0, ..., 1, 1, 1],
       [1, 1, 0, ..., 1, 1, 1],
       [1, 1, 0, ..., 1, 1, 1]])

In [25]:
df = pd.DataFrame(flips)

df

0  1  2  3  4  5  6  7
0     0  1  1  0  1  1  0  0
1     1  0  0  0  1  0  0  0
2     1  0  1  0  1  0  0  0
3     0  1  0  0  0  1  1  1
4     1  1  0  1  0  1  0  1
...  .. .. .. .. .. .. .. ..
9995  0  0  0  1  0  1  1  0
9996  0  0  0  0  1  1  0  0
9997  1  0  0  1  0  1  1  1
9998  1  1  0  0  1  1  1  1
9999  1  1  0  1  1  1  1  1

[10000 rows x 8 columns]

In [26]:
three_heads = df[(df.sum(axis=1) == 3)]

three_heads

0  1  2  3  4  5  6  7
2     1  0  1  0  1  0  0  0
7     1  1  0  0  0  0  0  1
12    1  0  0  0  0  1  1  0
19    1  0  0  1  1  0  0  0
20    0  1  0  1  0  1  0  0
...  .. .. .. .. .. .. .. ..
9963  0  0  1  0  0  1  1  0
9969  1  1  0  0  0  1  0  0
9977  1  1  0  0  0  0  0  1
9991  0  0  0  1  0  0  1  1
9995  0  0  0  1  0  1  1  0

[2167 rows x 8 columns]

In [28]:
len(three_heads)/n_cols

0.2167

### There is a 0.2167 (21.67%) probability of getting exactly 3 heads when flipping 8 coins

In [29]:
more_than_three = df[(df.sum(axis=1) > 3)]

more_than_three

0  1  2  3  4  5  6  7
0     0  1  1  0  1  1  0  0
3     0  1  0  0  0  1  1  1
4     1  1  0  1  0  1  0  1
5     1  0  1  1  1  1  0  0
6     1  1  1  1  1  1  0  1
...  .. .. .. .. .. .. .. ..
9993  1  0  1  0  1  1  0  1
9994  1  0  0  1  1  1  1  0
9997  1  0  0  1  0  1  1  1
9998  1  1  0  0  1  1  1  1
9999  1  1  0  1  1  1  1  1

[6359 rows x 8 columns]

In [30]:
len(more_than_three)/n_cols

0.6359

### There is a 0.6359 (63.59%) probability of getting more than 3 heads when flipping 8 coins.

##### 

##### 3. There are approximitely 3 web development cohorts for every 1 data science cohort at Codeup.
Assuming that Codeup randomly selects an alumni to put on a billboard, 
what are the odds that the two billboards I drive past both have data science students on them?


In [ ]:
# 3:1, web dev:ds
# 2 billboards


In [12]:
bb1_ds_prob = (1/4)
bb2_ds_prob = (1/4)

In [13]:
two_ds_bb = bb1_ds_prob * bb2_ds_prob

two_ds_bb

0.0625

### There is a 0.0625, or 6.25%, of both billboards being data science students

##### 

##### 4. Codeup students buy, on average, 3 poptart packages with a standard deviation of 1.5 a day from the snack vending machine. 
If on monday the machine is restocked with 17 poptart packages, how likely is it that I will be able to buy some poptarts on Friday afternoon? 

(Remember, if you have mean and standard deviation, use the np.random.normal) You'll need to make a  judgement call on how to handle some of your values


In [6]:
import statistics

In [7]:
mean = 3
std = 1.5

z_score = statistics.NormalDist(mean, std).zscore(5)

z_score

1.3333333333333333

In [12]:
df = np.random.normal(3, 1.5, 5)

In [8]:
import scipy

In [13]:
scipy.stats.mstats.zscore(df)

array([-0.83097528, -1.36942552,  1.33237461,  0.80964391,  0.05838228])

##### 5. Compare Heights
Men have an average height of 178 cm and standard deviation of 8cm.

Women have a mean of 170, sd = 6cm.

Since you have means and standard deviations, you can use np.random.normal to generate observations.

If a man and woman are chosen at random, what is the likelihood the woman is taller than the man?


In [19]:
women_heights = np.random.normal(170, 6, 10000)

men_heights = np.random.normal(178, 8, 10000)

In [20]:
women_heights > men_heights

array([False, False, False, ..., False, False, False])

In [21]:
(women_heights > men_heights).mean()

0.2082

##### There is a 0.2082 (20.82%) probability of women being taller then men

##### 

##### 6. When installing anaconda on a student's computer, there's a 1 in 250 chance that the download is corrupted and the installation fails. 

a. What are the odds that after having 50 students download anaconda, no one has an installation issue? 

b. 100 students?

c. What is the probability that we observe an installation issue within the first 150 students that download anaconda?

d. How likely is it that 450 students all download anaconda without an issue?


##### 

##### a. What are the odds that after having 50 students download anaconda, no one has an installation issue?

In [36]:
outcomes = ['s', 'f']

n_students = 50

In [37]:
downloads = np.random.choice(outcomes, size = (10000, n_students), p = [249/250, 1/250])

In [38]:
((downloads == 's').sum(axis = 1) == 50).mean()

0.8166

##### b. 100 students?

In [39]:
downloads = np.random.choice(outcomes, size = (10000, 100), p = [249/250, 1/250])

In [40]:
((downloads == 's').sum(axis = 1) == 50).mean()

0.0

##### c. What is the probability that we observe an installation issue within the first 150 students that download anaconda?

##### d. How likely is it that 450 students all download anaconda without an issue?

##### 7. There's a 70% chance on any given day that there will be at least one food truck at Travis Park. However, you haven't seen a food truck there in 3 days. 

How unlikely is this?

How likely is it that a food truck will show up sometime this week?


##### 8. If 23 people are in the same room, 

what are the odds that two of them share a birthday? 

What if it's 20 people? 40?
